File: ios/Releaf/Features/Notepad/Recents/Views/EarlierGrid.swift
import SwiftUI
import ReleafDesignSystem

/// 2-column grid of past pages — one card per `RecentsPage`. The
/// caller passes the list pre-sorted by `updatedAt` desc; the most-
/// recently-modified page is auto-promoted to a tall card spanning
/// two rows on the left, the rest flow as regular tiles in column-2
/// / paired rows below.
///
/// Empty days don't surface here anymore — the grid is a *page* feed,
/// so days with no captures simply contribute zero cards.
struct EarlierGrid: View {
    let pages: [RecentsPage]
    /// Tap a card to open that specific page. Default is no-op so
    /// the screen can drop the grid in without forcing the host to
    /// wire up navigation.
    var onOpenPage: (RecentsPage) -> Void = { _ in }

    var body: some View {
        if pages.isEmpty {
            EmptyView()
        } else {
            content
        }
    }

    private var content: some View {
        let featured = pages.first!
        let rest = Array(pages.dropFirst())
        // Right column hosts the next two pages, then any remaining
        // pages flow in 2-up rows below.
        let rightTop = rest.first
        let rightBottom = rest.dropFirst().first
        let below = rest.count > 2 ? Array(rest[2...]) : []

        return VStack(spacing: 12) {
            HStack(alignment: .top, spacing: 12) {
                // Tall featured (most-recently-modified page).
                Button { onOpenPage(featured) } label: {
                    PageCard(page: featured, variant: .tall)
                }
                .buttonStyle(.plain)
                .frame(maxWidth: .infinity)

                VStack(spacing: 12) {
                    if let p = rightTop {
                        Button { onOpenPage(p) } label: {
                            PageCard(page: p, variant: .regular)
                        }
                        .buttonStyle(.plain)
                    }
                    if let p = rightBottom {
                        Button { onOpenPage(p) } label: {
                            PageCard(page: p, variant: .regular)
                        }
                        .buttonStyle(.plain)
                    }
                }
                .frame(maxWidth: .infinity, alignment: .top)
            }
            // Remaining cards in 2-up pairs.
            ForEach(Array(stride(from: 0, to: below.count, by: 2)), id: \.self) { idx in
                HStack(alignment: .top, spacing: 12) {
                    Button { onOpenPage(below[idx]) } label: {
                        PageCard(page: below[idx], variant: .regular)
                    }
                    .buttonStyle(.plain)
                    .frame(maxWidth: .infinity)
                    if idx + 1 < below.count {
                        Button { onOpenPage(below[idx + 1]) } label: {
                            PageCard(page: below[idx + 1], variant: .regular)
                        }
                        .buttonStyle(.plain)
                        .frame(maxWidth: .infinity)
                    } else {
                        Color.clear.frame(maxWidth: .infinity)
                    }
                }
            }
        }
    }
}

// MARK: - PageCard

struct PageCard: View {
    enum Variant { case tall, regular }

    let page: RecentsPage
    let variant: Variant

    private static let dateFmt: DateFormatter = {
        let f = DateFormatter()
        f.locale = Locale(identifier: "en_US_POSIX")
        f.calendar = Calendar(identifier: .gregorian)
        f.timeZone = TimeZone(identifier: "UTC")
        f.dateFormat = "MMM d"
        return f
    }()

    private static let timeFmt: DateFormatter = {
        let f = DateFormatter()
        f.locale = Locale(identifier: "en_US_POSIX")
        f.calendar = Calendar(identifier: .gregorian)
        f.timeZone = TimeZone(identifier: "UTC")
        f.dateFormat = "h:mm a"
        return f
    }()

    var body: some View {
        switch variant {
        case .tall:    tallCard
        case .regular: regularCard
        }
    }

    // MARK: Tall

    private var tallCard: some View {
        VStack(alignment: .leading, spacing: 10) {
            dateLabel
            Text(page.title.isEmpty ? "untitled" : page.title)
                .font(Typography.h2Font)
                .foregroundColor(.textGreen)
                .lineLimit(2)
                .fixedSize(horizontal: false, vertical: true)

            if let description = description {
                Text(description)
                    .font(Typography.captionFont)
                    .foregroundColor(.textGreen)
                    .lineLimit(5)
                    .fixedSize(horizontal: false, vertical: true)
            }

            Spacer(minLength: 0)
            // Same surface order as the hero pip row, but coloured
            // for the leaf-on-leaf tall card. Each pill carries the
            // canonical `CaptureMode.systemIcon` plus the count, so
            // the user can see *what* the page contains and *how
            // much* at a glance — independent of the footer's
            // category + total tally below.
            earlierCapturePips
            footer(primaryColor: .textGreen, mutedColor: .textGreenMuted)
        }
        .padding(14)
        .frame(maxWidth: .infinity, alignment: .leading)
        .background(
            RoundedRectangle(cornerRadius: 20, style: .continuous)
                .fill(Color.bgFeatured)
        )
    }

    /// Capture-pip row used by the tall card. Mirrors the hero pip
    /// row's shape (icon + count, capsule bg) but with the leaf-on-
    /// leaf palette tuned for the tall card's `bgFeatured` (#DDEACD)
    /// surface. Pills with a zero count are skipped; if every count
    /// is zero the row collapses entirely so the spacing above the
    /// footer doesn't gain dead height.
    @ViewBuilder
    private var earlierCapturePips: some View {
        let counts = page.captureCounts
        if counts.total > 0 {
            HStack(spacing: 6) {
                if counts.photos    > 0 { earlierPip(CaptureMode.photos.systemIcon,   count: counts.photos) }
                if counts.scans     > 0 { earlierPip(CaptureMode.scans.systemIcon,    count: counts.scans) }
                if counts.voice     > 0 { earlierPip(CaptureMode.voice.systemIcon,    count: counts.voice) }
                if counts.todos     > 0 { earlierPip(CaptureMode.todo.systemIcon,     count: counts.todos) }
                if counts.contacts  > 0 { earlierPip(CaptureMode.contacts.systemIcon, count: counts.contacts) }
                if counts.locations > 0 { earlierPip(CaptureMode.location.systemIcon, count: counts.locations) }
                if counts.notes     > 0 { earlierPip("note.text",                     count: counts.notes) }
            }
            .padding(.bottom, 4)
        }
    }

    /// Slightly darker leaf tint than `bgFeatured` (#DDEACD) so the
    /// pill is legible without going loud. Deep-green icon matches
    /// the card's title typography; the count digit drops to
    /// `Color.green600` so the icon stays the primary read.
    private func earlierPip(_ systemName: String, count: Int) -> some View {
        HStack(spacing: 4) {
            Image(systemName: systemName)
                .font(.system(size: 11, weight: .medium))
                .foregroundColor(.green800)
            Text("\(count)")
                .font(.system(size: 11, weight: .medium))
                .foregroundColor(.green600)
        }
        .padding(.horizontal, 8)
        .padding(.vertical, 4)
        .background(Capsule().fill(Color.green200))
    }

    // MARK: Regular

    private var regularCard: some View {
        VStack(alignment: .leading, spacing: 8) {
            dateLabel
            Text(page.title.isEmpty ? "untitled" : page.title)
                .font(.system(size: 15, weight: .regular, design: .serif))
                .foregroundColor(.textPrimary)
                .lineLimit(2)
                .fixedSize(horizontal: false, vertical: true)
            if let description = description {
                Text(description)
                    .font(Typography.captionFont)
                    .foregroundColor(.textMuted)
                    .lineLimit(3)
                    .fixedSize(horizontal: false, vertical: true)
            }
            Spacer(minLength: 0)
            footer(primaryColor: .textGreen, mutedColor: .textMuted)
        }
        .padding(12)
        .frame(maxWidth: .infinity, alignment: .leading)
        .background(
            RoundedRectangle(cornerRadius: 16, style: .continuous)
                .fill(Color.bgSurface)
        )
        .overlay(
            RoundedRectangle(cornerRadius: 16, style: .continuous)
                .stroke(Color.borderFaint, lineWidth: 1)
        )
    }

    // MARK: Shared

    private var dateLabel: some View {
        Text(headerLabel.uppercased())
            .font(Typography.microWideFont)
            .kerning(1.4)
            .foregroundColor(.textGreenMuted)
    }

    private var headerLabel: String {
        let date = PageCard.dateFmt.string(from: page.createdAt)
        let time = PageCard.timeFmt.string(from: page.createdAt)
        return "\(date) · \(time)"
    }

    private var description: String? {
        let trimmed = page.description.trimmingCharacters(in: .whitespacesAndNewlines)
        return trimmed.isEmpty ? nil : trimmed
    }

    /// Bottom row: page category on the left, total capture count on
    /// the right. An amber dot precedes the category for imported
    /// pages. Both halves drop out gracefully when empty so the row
    /// never looks lopsided.
    @ViewBuilder
    private func footer(primaryColor: Color, mutedColor: Color) -> some View {
        let category = page.tags.first?.label
        let total = page.captureCounts.total
        if category != nil || total > 0 {
            HStack(spacing: 6) {
                if page.isImported {
                    Circle()
                        .fill(Color.accentImport)
                        .frame(width: 6, height: 6)
                }
                if let category {
                    Text(category.uppercased())
                        .font(Typography.microWideFont)
                        .kerning(1.2)
                        .foregroundColor(primaryColor)
                }
                Spacer(minLength: 0)
                if total > 0 {
                    Text("\(total) \(total == 1 ? "CAPTURE" : "CAPTURES")")
                        .font(Typography.microWideFont)
                        .kerning(1.2)
                        .foregroundColor(mutedColor)
                }
            }
        }
    }
}

#if DEBUG
#Preview {
    let pages = MockData.earlier
        .flatMap { $0.pages }
        .sorted { $0.updatedAt > $1.updatedAt }
    return EarlierGrid(pages: pages)
        .padding()
        .background(Color.bgCanvas)
}
#endif

File: apps/releaf/android/app/src/main/java/app/releaf/mobile/features/notepad/recents/ui/EarlierGrid.kt
package app.releaf.mobile.features.notepad.recents.ui

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.BasicText
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.outlined.EventNote
import androidx.compose.material3.Icon
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import app.releaf.mobile.features.notepad.recents.model.CaptureCounts
import app.releaf.mobile.features.notepad.recents.model.RecentsPage
import app.releaf.mobile.features.notepad.recents.model.isImported
import app.releaf.mobile.features.notepad.recents.theme.AccentImport
import app.releaf.mobile.features.notepad.recents.theme.BgFeatured
import app.releaf.mobile.features.notepad.recents.theme.BgSurface
import app.releaf.mobile.features.notepad.recents.theme.BorderFaint
import app.releaf.mobile.features.notepad.recents.theme.Green200
import app.releaf.mobile.features.notepad.recents.theme.Green600
import app.releaf.mobile.features.notepad.recents.theme.TextGreen
import app.releaf.mobile.features.notepad.recents.theme.TextGreenMuted
import app.releaf.mobile.features.notepad.recents.theme.TextMuted
import app.releaf.mobile.features.notepad.recents.theme.TextPrimary
import app.releaf.mobile.features.notepad.recents.theme.Type
import app.releaf.mobile.ui.components.CaptureMode
import java.time.format.DateTimeFormatter

/**
 * 2-column grid of past pages — one card per [RecentsPage], sorted
 * by `createdAt` desc by the caller. The most-recently-created page
 * is auto-promoted to a tall card spanning two rows on the left;
 * remaining cards flow as regular tiles in column-2 / paired rows
 * below.
 *
 * Empty days don't surface here anymore — the grid is a *page* feed,
 * so days with no captures simply contribute zero cards.
 */
@Composable
fun EarlierGrid(
    pages: List<RecentsPage>,
    onOpenPage: (RecentsPage) -> Unit,
    modifier: Modifier = Modifier,
) {
    if (pages.isEmpty()) return

    val featured = pages.first()
    val rest = pages.drop(1)

    // Right column hosts the next two pages, then any remaining pages
    // flow in 2-up rows below the featured tall card.
    val rightTop = rest.getOrNull(0)
    val rightBottom = rest.getOrNull(1)
    val below = if (rest.size > 2) rest.subList(2, rest.size) else emptyList()

    val rowHeight = 132.dp
    val tallHeight = rowHeight * 2 + 12.dp

    Column(
        modifier = modifier
            .fillMaxWidth()
            .padding(horizontal = 6.dp),
        verticalArrangement = Arrangement.spacedBy(12.dp),
    ) {
        Row(
            modifier = Modifier.fillMaxWidth(),
            horizontalArrangement = Arrangement.spacedBy(12.dp),
        ) {
            // Tall featured (most-recently-created page).
            Box(modifier = Modifier.weight(1f).height(tallHeight)) {
                TallPageCard(
                    page = featured,
                    onClick = { onOpenPage(featured) },
                    modifier = Modifier.fillMaxSize(),
                )
            }
            // Right column with two regular cards stacked.
            Column(
                modifier = Modifier.weight(1f),
                verticalArrangement = Arrangement.spacedBy(12.dp),
            ) {
                rightTop?.let {
                    RegularPageCard(
                        page = it,
                        onClick = { onOpenPage(it) },
                        modifier = Modifier.fillMaxWidth().height(rowHeight),
                    )
                }
                rightBottom?.let {
                    RegularPageCard(
                        page = it,
                        onClick = { onOpenPage(it) },
                        modifier = Modifier.fillMaxWidth().height(rowHeight),
                    )
                }
            }
        }
        // Remaining cards flow in 2-up pairs below.
        below.chunked(2).forEach { row ->
            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.spacedBy(12.dp),
            ) {
                row.forEach { p ->
                    Box(modifier = Modifier.weight(1f).height(rowHeight)) {
                        RegularPageCard(
                            page = p,
                            onClick = { onOpenPage(p) },
                            modifier = Modifier.fillMaxSize(),
                        )
                    }
                }
                if (row.size == 1) Spacer(Modifier.weight(1f))
            }
        }
    }
}

// ---------------------------------------------------------------------------
// Card variants
// ---------------------------------------------------------------------------

private val DATE_FMT = DateTimeFormatter.ofPattern("MMM d")
private val TIME_FMT = DateTimeFormatter.ofPattern("h:mm a")

@Composable
private fun TallPageCard(
    page: RecentsPage,
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
) {
    Box(
        modifier = modifier
            .clip(RoundedCornerShape(20.dp))
            .background(BgFeatured)
            .clickable(onClick = onClick)
            .padding(14.dp),
    ) {
        Column(modifier = Modifier.fillMaxSize()) {
            BasicText(
                text = headerLabel(page),
                style = Type.MicroLabel.copy(color = TextGreenMuted),
            )
            Spacer(Modifier.height(6.dp))
            BasicText(
                text = page.title.ifBlank { "untitled" },
                style = Type.CardTitle.copy(color = TextGreen, fontSize = 18.sp),
                maxLines = 2,
            )
            val description = page.description.takeIf { it.isNotBlank() }
            if (description != null) {
                Spacer(Modifier.height(8.dp))
                BasicText(
                    text = description,
                    style = Type.BodySmall.copy(color = TextGreen, fontSize = 12.sp),
                    maxLines = 5,
                )
            }
            Spacer(Modifier.weight(1f))
            // Same surface order as the hero pip row, but coloured
            // for the leaf-on-leaf tall card. Each pill carries the
            // canonical [CaptureMode] icon plus the count, so the
            // user can see *what* the page contains and *how much*
            // at a glance — independent of the footer's category +
            // total tally below.
            EarlierCapturePips(page.captureCounts)
            CardFooter(page = page, primaryColor = TextGreen, mutedColor = TextGreenMuted)
        }
    }
}

/**
 * Capture-pip row used by the EarlierGrid tall card. Mirrors the hero
 * pip row's shape (icon + count digit, capsule bg) but with the
 * leaf-on-leaf palette tuned for the tall card's `BgFeatured` (#DDEACD)
 * surface. Pills with a zero count are skipped; if every count is zero
 * the row collapses entirely so the spacing above the footer doesn't
 * gain dead height.
 */
@Composable
private fun EarlierCapturePips(counts: CaptureCounts) {
    if (counts.total <= 0) return
    Row(
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(6.dp),
        modifier = Modifier.padding(bottom = 8.dp),
    ) {
        if (counts.photos    > 0) EarlierPip(CaptureMode.Photos.icon,    counts.photos)
        if (counts.scans     > 0) EarlierPip(CaptureMode.Scans.icon,     counts.scans)
        if (counts.voice     > 0) EarlierPip(CaptureMode.Voice.icon,     counts.voice)
        if (counts.todos     > 0) EarlierPip(CaptureMode.Todo.icon,      counts.todos)
        if (counts.contacts  > 0) EarlierPip(CaptureMode.Contacts.icon,  counts.contacts)
        if (counts.locations > 0) EarlierPip(CaptureMode.Location.icon,  counts.locations)
        if (counts.notes     > 0) EarlierPip(Icons.AutoMirrored.Outlined.EventNote, counts.notes)
    }
}

@Composable
private fun EarlierPip(icon: ImageVector, count: Int) {
    // Slightly darker leaf tint than the card's `BgFeatured` (#DDEACD)
    // so the pill is legible without going loud. Deep-green icon
    // matches the card's title typography; the count digit drops to
    // [Green600] so the icon stays the primary read.
    Row(
        modifier = Modifier
            .clip(RoundedCornerShape(999.dp))
            .background(Green200)
            .padding(horizontal = 8.dp, vertical = 4.dp),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(4.dp),
    ) {
        Icon(
            imageVector = icon,
            contentDescription = null,
            tint = TextGreen,
            modifier = Modifier.size(13.dp),
        )
        BasicText(
            text = count.toString(),
            style = Type.BodySmall.copy(
                color = Green600,
                fontSize = 10.sp,
                fontWeight = androidx.compose.ui.text.font.FontWeight.Medium,
            ),
        )
    }
}

@Composable
private fun RegularPageCard(
    page: RecentsPage,
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
) {
    Column(
        modifier = modifier
            .clip(RoundedCornerShape(16.dp))
            .background(BgSurface)
            .border(1.dp, BorderFaint, RoundedCornerShape(16.dp))
            .clickable(onClick = onClick)
            .padding(12.dp),
    ) {
        BasicText(
            text = headerLabel(page),
            style = Type.MicroLabel.copy(color = TextGreenMuted),
        )
        Spacer(Modifier.height(6.dp))
        BasicText(
            text = page.title.ifBlank { "untitled" },
            style = Type.CardTitle.copy(color = TextPrimary, fontSize = 15.sp),
            maxLines = 2,
        )
        val description = page.description.takeIf { it.isNotBlank() }
        if (description != null) {
            Spacer(Modifier.height(6.dp))
            BasicText(
                text = description,
                style = Type.BodySmall.copy(color = TextMuted, fontSize = 12.sp),
                maxLines = 3,
            )
        }
        Spacer(Modifier.weight(1f))
        CardFooter(page = page, primaryColor = TextGreen, mutedColor = TextMuted)
    }
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

/** "APR 25 · 8:05 AM" header that mirrors the hero's eyebrow shape. */
private fun headerLabel(page: RecentsPage): String =
    "${page.createdAt.format(DATE_FMT)} · ${page.createdAt.format(TIME_FMT)}"
        .uppercase()

/** Bottom row: page category on the left, total capture count on
 *  the right. The amber dot for imported pages stays as a marker on
 *  the left of the category. Both sides drop out gracefully when
 *  empty so the row never looks lopsided. */
@Composable
private fun CardFooter(
    page: RecentsPage,
    primaryColor: androidx.compose.ui.graphics.Color,
    mutedColor: androidx.compose.ui.graphics.Color,
) {
    val category = page.tags.firstOrNull()?.display
    val total = page.captureCounts.total
    val showAny = category != null || total > 0
    if (!showAny) return
    Row(
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(6.dp),
        modifier = Modifier.fillMaxWidth(),
    ) {
        if (page.isImported()) {
            Box(
                modifier = Modifier
                    .size(6.dp)
                    .clip(CircleShape)
                    .background(AccentImport),
            )
        }
        if (category != null) {
            BasicText(
                text = category.uppercase(),
                style = Type.MicroLabel.copy(color = primaryColor, fontSize = 9.sp),
            )
        }
        Spacer(Modifier.weight(1f))
        if (total > 0) {
            BasicText(
                text = "$total ${if (total == 1) "capture" else "captures"}".uppercase(),
                style = Type.MicroLabel.copy(color = mutedColor, fontSize = 9.sp),
            )
        }
    }
}
